### PR TITLE
GC: Fix survival rates for LOH/POH

### PR DESCRIPTION
**Pull request by markples**
The survival rate calculation for BGCs includes objects allocated during the BGC in the numerator but not denominator.  This fixes that by adding them to size_before.  The existing counters are reset mid-BGC, so the mark and sweep phases need to be added at separate points.  I verified the fix over several gcperfsim runs and watched the counters getting updated in the debugger.

This includes some new and moved comments from going through the phases of BGC and putting similarly ifdef-ed code together.

Partial fix for #100594

### CODE REVIEW

File: src/coreclr/gc/gc.cpp [hunk 1]
```diff
@@ -46221,6 +46237,7 @@ void gc_heap::background_sweep()
             concurrent_print_time_delta ("Swe SOH");
             FIRE_EVENT(BGC1stSweepEnd, 0);
 
+            //block concurrent allocation for large objects
```
Comment (Maoni0) [inline src/coreclr/gc/gc.cpp:46240]: >large [](http://example.com/codeflow?start=46&length=5)

UOH objects

File: src/coreclr/gc/gcpriv.h [hunk 2]
```diff
@@ -3436,6 +3439,13 @@ class gc_heap
 
     PER_HEAP_FIELD_SINGLE_GC uint8_t* next_sweep_obj;
     PER_HEAP_FIELD_SINGLE_GC uint8_t* current_sweep_pos;
+
+    PER_HEAP_FIELD_SINGLE_GC uint64_t loh_a_no_bgc;
+    PER_HEAP_FIELD_SINGLE_GC uint64_t loh_a_bgc_marking;
+    PER_HEAP_FIELD_SINGLE_GC uint64_t loh_a_bgc_planning;
+    PER_HEAP_FIELD_SINGLE_GC uint64_t poh_a_bgc_marking;
+    PER_HEAP_FIELD_SINGLE_GC uint64_t poh_a_bgc_planning;
```
Comment (Maoni0) [inline src/coreclr/gc/gcpriv.h:3447]: perhaps have a little array to store both loh and poh data?

File: src/coreclr/gc/gc.h [hunk 3]
```diff
@@ -124,7 +124,10 @@ enum gc_generation_num
     ephemeral_generation_count = max_generation,
 
     // number of all generations
-    total_generation_count = poh_generation + 1
+    total_generation_count = poh_generation + 1,
+
+    // number of uoh generations
+    uoh_generation_count = total_generation_count - ephemeral_generation_count
 };
```
Comment (Maoni0) [inline src/coreclr/gc/gc.h:131]: this should be 2, not 3, ie,

`total_generation_count - uoh_start_generation`

Comment (markples) [reply src/coreclr/gc/gc.h:131]: wow, I must have read that as "loh_generation_count" which of course is different.  fixed.

**Review by Maoni0 — APPROVED** [review]

**Pull request merged**